src/cascade/input_data/db/bundle.py

```diff
@@ -0,0 +1,327 @@
+
+
+def _upload_study_covariates_to_tier_3(cursor, model_version_id, covariate_data):
+    """Updloads study covariate mappings to tier 3 attached to the current model_version_id.
+    """
+
+    insert_query = f"""
+    INSERT INTO epi.t3_model_version_study_covariate (
+        model_version_id,
+        bundle_id,
+        seq,
+        study_covariate_id
+    ) VALUES (
+        {model_version_id}, %s, %s, %s
+    )
+    """
+
+    cursor.executemany(insert_query, covariate_data.values())
+
+    CODELOG.debug(f"uploaded {len(covariate_data)} lines of covariate")
+
+
+def freeze_bundle(execution_context) -> bool:
+    """Freezes the bundle data attached to the current model_version_id if necessary.
+
+    The freezing process works as follows:
+    If there is any data in the tier 3 table for the model_version_id then it is assumed to already be frozen.
+    Otherwise download all data for the bundle from the tier 2 table in the main epi database and reupload it
+    to the dismodAT database (ie. promote it).
+
+    Returns:
+      True if any data was promoted and False otherwise
+    """
+
+    model_version_id = execution_context.parameters.model_version_id
+
+    database = execution_context.parameters.database
+    if _bundle_is_frozen(execution_context):
+        CODELOG.info(
+            f"Bundle data for model_version_id {model_version_id} on '{database}' already frozen, doing nothing."
+        )
+        return False
+    else:
+        CODELOG.info(f"Freezing bundle data for model_version_id {model_version_id} on '{database}'")
+        bundle_id = _get_bundle_id(execution_context)
+        bundle_data = _get_bundle_data(execution_context, bundle_id, tier=2)
+        covariate_data = _get_study_covariates(execution_context, bundle_id, tier=2)
+        with cursor(execution_context) as c:
+            _upload_bundle_data_to_tier_3(c, model_version_id, bundle_data)
+            _upload_study_covariates_to_tier_3(c, model_version_id, covariate_data)
+        return True
+
+
+def _normalize_measures(data):
+    """Transform measure_ids into canonical measure names
+    """
+    data = data.copy()
+    data["measure"] = data.measure_id.apply(MEASURES.get)
+    return data
+
+
+def _normalize_sex(data):
+    """Transform sex_ids into strings
+    """
+    data = data.copy()
+    data["sex"] = data.sex_id.apply({1: "Male", 2: "Female", 3: "Both"}.get)
+    return data
+
+
+def _normalize_bundle_data(data):
+    """Normalize bundle columns, strip extra columns and index on `seq`
+    """
+    data = _normalize_measures(data)
+    data = _normalize_sex(data)
+
+    data = data.set_index("seq")
+
+    cols = ["measure", "mean", "sex", "standard_error", "age_start", "age_end", "year_start", "year_end"]
+
+    return data[cols]
+
+
+def _covariate_ids_to_names(execution_context, study_covariates):
+    """Convert study_covariate_ids to canonical study covariate names
+    """
+    query = """
+    select study_covariate_id, study_covariate
```

Comment:
Should we always capitalize these; i.e. SELECT?

Reply:
I don't want to. We can if it really bothers people but it's totally unnecessary and ugly.